onnxruntime/core/session/environment.cc

```diff
@@ -68,6 +72,11 @@ Internal copy node
 Internal copy node
 )DOC");
 
+	DataTransferManager::Instance().RegisterDataTransfer(std::make_unique<CPUDataTransfer>());
+#ifdef USE_CUDA
+    DataTransferManager::Instance().RegisterDataTransfer(std::make_unique<GPUDataTransfer>());    
```

Comment:
Like we discussed offline, I think the registration of the specific copy functions should be done by the providers as they own the copy logic.

Reply:
think it twice, I'm moving the registration into session level so that it can be registered when initializing an inference session.

Reply:
I don't think the question is whether it's done in session or environment. Conceptually it's the EPs that provide the copy function and they should be the ones doing the registration. The session should only care about registering the provided EPs.

Reply:
I would have thought that conceptually the copy logic is device specific. Whilst a given EP may have a strong affinity to a device, it isn't necessarily the only EP that uses the device. To me that would suggest the copy logic exists outside of a specific EP, but the EP would ensure the necessary copy functions are registered for the devices it uses.

Reply:
Yes, the idea is that an EP (and not the session or the environment) register the necessary copy functions since it's the one that is managing a set of devices. However, this requires either the DataTransferMgr to be a singleton or it's instance be passed to the EP somehow. Ke and I were discussing this yesterday; didn't reach a consensus yet.